hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java

```diff
@@ -63,34 +71,69 @@ public UniqueTokenStore(final EntityIdSource ids,
 	) {
 		super(ids, validator, properties, tokens, tokenRelsLedger);
 		this.uniqueTokenSupplier = uniqueTokenSupplier;
+		this.provisionalUniqueTokens = new ArrayList<>(); // TODO MAX_SIZE
+		this.lastMintedSerialNumbers = new ArrayList<>();
 	}
 
 	@Override
-	public ResponseCodeEnum mint(final TokenID tId, final String memo, final RichInstant creationTime) {
-		return tokenSanityCheck(tId, (merkleToken -> {
+	public List<Long> popLastMintedSerialNumbers() {
+		var copy = new ArrayList<>(lastMintedSerialNumbers);
+		lastMintedSerialNumbers.clear();
+		return copy;
+	}
+
+	@Override
+	public ResponseCodeEnum wipe(final AccountID aId, final TokenID tId, final long wipingAmount, final boolean skipKeyCheck) {
+		return null;
+	}
+
+	@Override
+	public ResponseCodeEnum mint(final TokenMintTransactionBody txBody, final RichInstant creationTime) {
+		return tokenSanityCheck(txBody.getToken(), merkleToken -> {
 			if (!merkleToken.hasSupplyKey()) {
 				return TOKEN_HAS_NO_SUPPLY_KEY;
 			}
-			var mintResult = super.mint(tId, 1);
-			if (!mintResult.equals(OK)) {
-				return mintResult;
-			}
-			final var suppliedTokens = uniqueTokenSupplier.get();
-			final var eId = EntityId.fromGrpcTokenId(tId);
+			final var eId = EntityId.fromGrpcTokenId(txBody.getToken());
 			final var owner = merkleToken.treasury();
-			final long serialNum = merkleToken.incrementSerialNum();
-
-			final var nftId = new MerkleUniqueTokenId(eId, serialNum);
-			final var nft = new MerkleUniqueToken(owner, memo, creationTime);
-			suppliedTokens.put(nftId, nft);
-			return OK;
-		}));
+			final var metadataList = txBody.getMetadataList();
+			final AtomicLong serialNum = new AtomicLong(merkleToken.getCurrentSerialNum()); // create them without incrementing serialNum
+			metadataList.stream().map(el -> {
+				String metaAsStr = el.toStringUtf8();
+				final var nftId = new MerkleUniqueTokenId(eId, serialNum.get());
+				final var nft = new MerkleUniqueToken(owner, metaAsStr, creationTime);
+				serialNum.getAndIncrement();
+				return Pair.of(nftId, nft);
+			}).forEach(provisionalUniqueTokens::add);
 
+			if (!checkProvisional()) {
+				provisionalUniqueTokens.clear();
+				return INVALID_TRANSACTION_BODY;
+			}
+			var adjustmentResult = tryAdjustment(merkleToken.treasury().toGrpcAccountId(), txBody.getToken(), provisionalUniqueTokens.size());
```

Comment:
We have to change the `serialNumber` property in the Merkle Token as well

Reply:
This is happening in the commitProvisional method right now